### PR TITLE
Add opentelemetry source code attributes to spans

### PR DESCRIPTION
**Pull request by lilymara-onesignal**
## Motivation

There is some more context in #1397 but the main idea is that we should be able to trace a Span back to the location in the code where it was created. Currently many crates have spans with very generic names (`h2` has `send_data` and `reserve_capacity`) and this makes it very difficult to figure out where spans are coming from. This change will make it much easier to find the location in the code where spans are created.

## Solution

We can capture the location information using the existing location data on the `tracing::Span`.

The opentelemetry specification calls for a number of attributes to correlate
traces to their location in the code. They are documented here [1]. This commit
adds support for the following fields based on the tracing span metadata (all
relative to span creation):

- `code.namespace`: Crate & module path (`my_crate::my_mod`)
- `code.filepath`: Relative path to the source code file (`src/my_mod.rs`)
- `code.lineno`: Line number in the file indicated by `code.filepath` (`72`)

As written this will annotate all spans with these attributes. If we want to be
a bit more conservative, I could add an instance variable to the Subscriber that
allows users to opt-in or opt-out of this functionality.

[1]: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/span-general.md#source-code-attributes

# Note

I saw that there is a requirement to add tests for new functionality, but I do not see any tests around the attributes that this crate adds to opentelemetry Spans and Events. I have not added any tests because I'm not sure how to write those or where they would go.

### CODE REVIEW

File: tracing-opentelemetry/src/subscriber.rs [hunk 1]
```diff
@@ -421,6 +421,22 @@ where
...
+            builder_attrs.push(KeyValue::new("code.filepath", filename));
+        }
+
+        if let Some(module) = meta.module_path() {
+            builder_attrs.push(KeyValue::new("code.namespace", module));
+        }
+
+        if let Some(line) = meta.line() {
+            builder_attrs.push(KeyValue::new("code.lineno", line as i64));
+        }
```
Comment (hawkw) [inline tracing-opentelemetry/src/subscriber.rs:438]: i wonder if we want to only produce this attribute if the `module_path` or `file` values are present? a line number without a file or module name doesn't make a lot of sense.

on the other hand, the `tracing` macros will never generate metadata that has a line number without a filename or module path, and i'd be very surprised if any other instrumentation did that. so, it's not a big deal either way.

**Review by hawkw — APPROVED** [review]
looks good to me!

File: tracing-opentelemetry/src/subscriber.rs (hunk 1 above)
Comment (lilymara-onesignal) [reply tracing-opentelemetry/src/subscriber.rs:438]: My thinking was that we should try to be defensive and include the maximum possible amount of information on each span.

**Comment by jtescher** [comment]
Looking at the existing benchmarks it appears this may have a performance impact for cases that don't have many fields. Perhaps worth adding a flag here?

**Comment by altanozlu** [comment]
Is there any way to disable it ?

**Pull request merged**